cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -202,8 +209,138 @@ impl GrammarAST {
                 spans: vec![Span::new(0, 0)],
             });
         }
+
+        for sym in &self.expect_unused {
+            match sym {
+                crate::yacc::ast::Symbol::Rule(sym_name, sym_span) => {
+                    if self.get_rule(sym_name).is_none() {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownRuleRef(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+                crate::yacc::ast::Symbol::Token(sym_name, sym_span) => {
+                    if !self.has_token(sym_name) {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownToken(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+            }
+        }
         Ok(())
     }
+
+    pub(crate) fn unused_symbol_warnings(&self) -> impl Iterator<Item = YaccGrammarWarning> {
```

Comment:
I'm a bit slow, so I didn't realise at first quite what this is doing. Basically it's a reachability algorithm. I *think* we can simplify it a bit. In pseudocode something like this might work?

```rust
let expected_unused_rules = HashSet()...;
let expected_unused_tokens = HashSet()...;

let mut todo = vec![start_rule];
let mut seen_rules = HashSet::new();
let mut seen_tokens = HashSet::new();
while !todo.is_empty() {
    let rule = todo.pop();
    for sym in rule {
    	match sym {
	    Rule(x) => {
	       if !seen_rules.insert(x) {
	           todo.push(x);
	       }
	    }
	    Token(x) => seen_tokens.insert(x)
	}
    }
}

let unexpected_unused_rules = all_rules.difference(seen_rules).difference(expected_unused_rules);
let unexpected_unused_tokens = all_tokens.difference(seen_tokens).difference(expected_unused_tokens);
```

Reply:
I'll give it a try, but I think part of what is making my impl a bit ugly, is the need to have `Span` in the return value,
these can't go in the `HashSet`'s .

So my `all_rules` and `all_tokens` equivalents have a different type than `seen_rules` and `expect_unused_rules` equivalents.  So for this to work I think i'd have add an extra pass over `unexpected_unused_*` to get spans.

Reply:
So I did a start at using separate rules/tokens sets in [49645ce](https://github.com/softdevteam/grmtools/pull/345/commits/49645ce2e0a0a61233f625a38ee0bff7b05503d5). 

I'll have to look into whether it will be possible to use something like `difference` later though.